### PR TITLE
Check whether the task finishes before deferring the task for SageMakerTransformOperatorAsync

### PR DESCRIPTION
**Pull request by Lee-W**
Like the [SnowflakeOperatorAsync](https://github.com/astronomer/astronomer-providers/blob/b2dade827ad8425952b2f5313b6a2863cb0c3e5e/astronomer/providers/snowflake/operators/snowflake.py#L217), we try to verify if the submitted job has already completed before deferring it to prevent unnecessary deferring. This way, we can skip deferring the task if it has already been finished.

### CODE REVIEW

**Comment by Lee-W** [comment]
圖片

**Comment by Lee-W** [comment]
圖片

File: astronomer/providers/amazon/aws/operators/sagemaker.py [hunk 1]
```diff
@@ -179,24 +181,40 @@ def execute(self, context: Context) -> None:  # type: ignore[override]
...
-                    job_name=transform_config["TransformJobName"],
-                    job_type="Transform",
-                    response_key="TransformJobStatus",
-                ),
-                method_name="execute_complete",
-            )
 
-    def execute_complete(self, context: "Context", event: Dict[str, Any]) -> Dict[str, Any]:
+        response = self.hook.describe_transform_job(transform_config["TransformJobName"])
+        status = response["TrainingJobStatus"]
```
Comment (pankajkoti) [inline astronomer/providers/amazon/aws/operators/sagemaker.py:186]:
```suggestion
        status = response["TransformJobStatus"]
```

File: astronomer/providers/amazon/aws/operators/sagemaker.py [hunk 2]
```diff
@@ -179,24 +181,40 @@ def execute(self, context: Context) -> None:  # type: ignore[override]
...
+        if status in self.hook.failed_states:
+            raise AirflowException(f"SageMaker job failed because {response['FailureReason']}")
+
+        if status == "Completed":
+            self.log.info(f"{self.task_id} completed successfully.")
+            transform_config = self.config.get("Transform", self.config)
```
Comment (pankajkoti) [inline astronomer/providers/amazon/aws/operators/sagemaker.py:192]: Do we need this line to get `transform_config`, we already have this somewhere above, no?

File: tests/amazon/aws/operators/test_sagemaker.py [hunk 3]
```diff
@@ -219,14 +219,91 @@ class TestSagemakerTransformOperatorAsync:
     CHECK_INTERVAL = 5
     MAX_INGESTION_TIME = 60 * 60 * 24 * 7
 
+    @mock.patch("astronomer.providers.amazon.aws.operators.sagemaker.SageMakerTransformOperatorAsync.defer")
+    @mock.patch.object(SageMakerHook, "describe_model", return_value={})
+    @mock.patch.object(
+        SageMakerHook,
+        "describe_transform_job",
+        return_value={"TrainingJobStatus": "Failed", "FailureReason": "it just failed"},
```
Comment (pankajkoti) [inline tests/amazon/aws/operators/test_sagemaker.py:227]:
```suggestion
        return_value={"TransformJobStatus": "Failed", "FailureReason": "it just failed"},
```

File: tests/amazon/aws/operators/test_sagemaker.py [hunk 4]
```diff
@@ -219,14 +219,91 @@ class TestSagemakerTransformOperatorAsync:
...
+            max_ingestion_time=self.MAX_INGESTION_TIME,
+        )
+        with pytest.raises(AirflowException):
+            task.execute(context)
+        assert not mock_defer.called
+
+    @mock.patch("astronomer.providers.amazon.aws.operators.sagemaker.SageMakerTransformOperatorAsync.defer")
+    @mock.patch.object(SageMakerHook, "describe_model", return_value={})
+    @mock.patch.object(
+        SageMakerHook, "describe_transform_job", return_value={"TrainingJobStatus": "Completed"}
```
Comment (pankajkoti) [inline tests/amazon/aws/operators/test_sagemaker.py:260]:
```suggestion
        SageMakerHook, "describe_transform_job", return_value={"TransformJobStatus": "Completed"}
```

File: tests/amazon/aws/operators/test_sagemaker.py [hunk 5]
```diff
@@ -219,14 +219,91 @@ class TestSagemakerTransformOperatorAsync:
...
+            task_id=self.TASK_ID,
+            check_if_job_exists=False,
+            check_interval=self.CHECK_INTERVAL,
+            max_ingestion_time=self.MAX_INGESTION_TIME,
+        )
+        task.execute(context)
+        assert not mock_defer.called
+
+    @mock.patch.object(
+        SageMakerHook, "describe_transform_job", return_value={"TrainingJobStatus": "InProgress"}
```
Comment (pankajkoti) [inline tests/amazon/aws/operators/test_sagemaker.py:290]:
```suggestion
        SageMakerHook, "describe_transform_job", return_value={"TransformJobStatus": "InProgress"}
```

File: astronomer/providers/amazon/aws/operators/sagemaker.py (hunk 2 above)
Comment (Lee-W) [reply astronomer/providers/amazon/aws/operators/sagemaker.py]: Yep, miss that one. Just removed it. Thanks!

**Review by pankajastro — APPROVED** [review]
LGTM

**Pull request merged**